### PR TITLE
[processor/filter] Add ottl to filter processor

### LINKED ISSUE #5679: Add ability to filter log records by body (maybe expr language)
**Is your feature request related to a problem? Please describe.**

I want to filter in/out logs by specific body content

**Describe the solution you'd like**
I can see two solutions. All of them should be implemented in [filterprocessor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/0ba52a57f2345f13c16715418108aa8c57019e1f/processor/filterprocessor)

 - Simple solution: run regex against stringified (or only for string typed) body
 - Complex solution: add support for [expr language](https://github.com/antonmedv/expr/blob/master/docs/Language-Definition.md). This is already done for metrics 

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### LINKED ISSUE #7151: New component: Span Event Filter Processor
**Is your feature request related to a problem? Please describe.**

While [span events](https://opentelemetry.io/docs/instrumentation/go/manual/#events) can be useful in some circumstances, other times they can be noisy or undesired. Some things like service meshes (Envoy, Istio) and auto-instrumentation (gRPC) attach a lot of span events, and there is currently no way to filter them out of traces.

**Describe the solution you'd like**

A Span Event Filter Processor would provide the ability to filter out span events based on attributes and names.

**Describe alternatives you've considered**

Attempting to drop span events from all possible instrumentations is not feasible. Since it's also not possible (yet?) to drop individual spans, dropping the trace entirely or sampling more heavily are other options... but that results in losing more valuable data to filter out unwanted data. 

**Additional context**
​
There is already a [(now closed) PR ](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/5716) that I would recommend reviving for this purpose.

### LINKED ISSUE #13578: [processor/transform] Add `drop` action support
`drop` action is pretty important functionality to replace filter processors. It's already being mentioned in the docs but not implemented yet.

Subtasks per data type:

- [ ] #13579
- [ ] #13580
- [ ] #13581

### LINKED ISSUE #13579: [pkg/telemetryquerylanguage] Add `drop` action support for metrics

### LINKED ISSUE #13580: [pkg/telemetryquerylanguage] Add `drop` action support for traces

### LINKED ISSUE #13581: [pkg/telemetryquerylanguage] Add `drop` action support for logs

### PR DESCRIPTION
**Pull request by TylerHelmuth**
**Description:**
Adds a new configuration options to the filterprocessor that allow filtering spans, spanevents, metrics, datapoints, and logs using OTTL conditions.

**Link to tracking Issue:**
Closes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/7151
Closes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/13578
Closes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/13579
Closes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/13580
Closes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/13581
Closes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/5679

**Testing:** 
Unit tests

**Documentation:**
Updated README

### CODE REVIEW

File: processor/filterprocessor/config.go [hunk 1]
```diff
@@ -39,8 +48,30 @@ type Config struct {
...
+}
+
+type Conditions struct {
+	Resource  string `mapstructure:"resource"`
+	Scope     string `mapstructure:"scope"`
+	Span      string `mapstructure:"span"`
+	SpanEvent string `mapstructure:"spanevent"`
+	Metric    string `mapstructure:"metric"`
+	DataPoint string `mapstructure:"datapoint"`
+	Log       string `mapstructure:"log"`
```
Comment (bogdandrutu) [inline processor/filterprocessor/config.go:62]: I think we should not offer resource and scope, since it is not clear what dropping a "resource" mean. e.g. does it mean dropping the resource or the resource and all the data associated with it? 

Also would start with only supporting logrecord/metric/span initially, then we can extend.

File: processor/filterprocessor/config.go [hunk 2]
```diff
@@ -39,8 +48,30 @@ type Config struct {
 	Logs LogFilters `mapstructure:"logs"`
 
 	Spans SpanFilters `mapstructure:"spans"`
+
+	Conditions Conditions `mapstructure:"conditions"`
+}
+
+type Conditions struct {
```
Comment (bogdandrutu) [inline processor/filterprocessor/config.go:55]: Somewhere there is a need to have a "list" of conditions, since I want to be able to remove spans based on more conditions.

Comment (TylerHelmuth) [reply processor/filterprocessor/config.go:55]: In OTTL this means AND or OR statements all within the same condition statement

Comment (bogdandrutu) [reply processor/filterprocessor/config.go]: The conditions may get super complicated (for OR case especially, if I want to remove multiple metrics), I would prefer to have multiple simpler conditions, and if any of them matches then the telemetry is removed.

Comment (bogdandrutu) [reply processor/filterprocessor/config.go]: Essentially if I have 3 complete independent conditions this force me to do a complicated 1 condition:
* Filter "healthcheck" spans
* Filter spans from service "envelope"
* Filter spans from pod "foo"

Will get super complicated.

File: processor/filterprocessor/config.go [hunk 3]
```diff
@@ -68,6 +85,16 @@ type SpanFilters struct {
...
+
+	// Span is an OTTL condition for an ottlspan context.
+	// If the condition resolves to true, the span will be dropped.
+	// Supports `and`, `or`, and `()`
+	Span string `mapstructure:"span"`
+
+	// SpanEvent is an OTTL condition for an ottlspanevent context.
+	// If the condition resolves to true, the span event will be dropped.
+	// Supports `and`, `or`, and `()`
+	SpanEvent string `mapstructure:"spanevent"`
```
Comment (bogdandrutu) [inline processor/filterprocessor/config.go:97]: Should these be under a "traces" entry (not spans) to be consistent with logs/metrics which are the name of the signals.

Comment (bogdandrutu) [reply processor/filterprocessor/config.go:97]: >I could create the traces section now and only include the new config options, or we could do a breaking change at a later time.

Yes only the new config should be there I think, the old one we are planning to deprecate anyway.

Comment (TylerHelmuth) [reply processor/filterprocessor/config.go]: Done

File: processor/filterprocessor/config.go (hunk 2 above)
Comment (TylerHelmuth) [reply processor/filterprocessor/config.go]: Ok, added multiple condition support.

File: processor/filterprocessor/config.go [hunk 4]
```diff
@@ -80,6 +112,11 @@ type LogFilters struct {
 	// all other logs should be included.
 	// If both Include and Exclude are specified, Include filtering occurs first.
 	Exclude *LogMatchProperties `mapstructure:"exclude"`
+
+	// LogConditions is a list of OTTL conditions for an ottllog context.
+	// If any condition resolves to true, the log event will be dropped.
+	// Supports `and`, `or`, and `()`
+	LogConditions []string `mapstructure:"log_conditions"`
```
Comment (bogdandrutu) [inline processor/filterprocessor/config.go:119]: log_record?

File: processor/filterprocessor/config.go [hunk 5]
```diff
@@ -55,6 +64,16 @@ type MetricFilters struct {
 
 	// RegexpConfig specifies options for the Regexp match type
 	RegexpConfig *regexp.Config `mapstructure:"regexp"`
+
+	// MetricConditions is a list of OTTL conditions for an ottlmetric context.
+	// If any condition resolves to true, the metric will be dropped.
+	// Supports `and`, `or`, and `()`
+	MetricConditions []string `mapstructure:"metric_conditions"`
```
Comment (bogdandrutu) [inline processor/filterprocessor/config.go:71]: Do we need the "_conditions" suffix? I am not convinced about it. Happy to hear opinions.

Comment (TylerHelmuth) [reply processor/filterprocessor/config.go:71]: I added `_conditions` because I wanted to be explicit in the configuration that these were all the OTTL conditions.  In  the yaml I think it looks good 

```yaml
filter/ottl:
  traces:
    span:
      - 'attributes["test"] == "pass"'
    spanevent:
      - 'attributes["test"] == "pass"'
  metrics:
    metric:
      - 'name == "pass"'
    datapoint:
      - 'attributes["test"] == "pass"'
  logs:
    log:
      - 'attributes["test"] == "pass"'
```
but I was worried about new and existing users trying to understand the configuration option.  Adding `_condition` was an attempt to help with that documentation.

File: processor/filterprocessor/logs.go [hunk 6]
```diff
@@ -18,21 +18,41 @@ import (
...
 type filterLogProcessor struct {
 	cfg            *Config
 	excludeMatcher filterlog.Matcher
 	includeMatcher filterlog.Matcher
 	logger         *zap.Logger
+	logConditions  []*ottl.Statement[ottllog.TransformContext]
 }
 
 func newFilterLogsProcessor(logger *zap.Logger, cfg *Config) (*filterLogProcessor, error) {
+	if cfg.Logs.LogConditions != nil {
```
Comment (bogdandrutu) [inline processor/filterprocessor/logs.go:42]: Add a warning if using the old config? And mark the old one as deprecated?

If you want you can do a separate PR.

Comment (TylerHelmuth) [reply processor/filterprocessor/logs.go:42]: I still need to check that OTTL has 100% of the features offered by the existing configuration options so I'd opt not to add any Deprecated comments/warnings yet.

**Review by bogdandrutu — APPROVED** [review]
